### ISSUE TITLE
Re-implement Skia versions

### ISSUE THREAD

**Issue by azeno** [vvvv/VL.MediaFoundation#2]
Currently VideoIn and VideoPlayer output Stride textures to make full use of hardware acceleration.

If we bring back the version returning images the hardware acceleration doesn't make much sense as it requires a download from the GPU. So instead we could work with a non-accelerated pipeline in those cases.

**Comment by hameleon-ed** [comment]
Currently, the VideoPlayer node in VL output correctly only horizontal video like 1920x1080. Instead of that square videos or vertical videos output as like video lines shifted. I tried several not horizontal videos and all of them shown incorrect texture.

**Comment by YanYas** [comment]
I agree. on the projects I've worked on with this library Stride hasn't been required and the results were using a Skia renderer.

You may have to split the repo in two where the core MediaFoundation is a module for the different output repos.

**Comment by azeno** [comment]
Now that we use ANGLE as rendering back-end for Skia, we can create Skia images out of the given Direct3D11 textures directly -> no GPU download -> same performance as in Stride. So the only remaining question would be how to organize the packages. 

###  One package
We deliver one package called `VL.Video.MediaFoundation` containing two nodes `VideoPlayer (Stride)` and `VideoPlayer (Skia)`.

`+` Only one package. Easier to install.
`-` For an export to work, the package would need to reference both Skia and Stride increasing the output size.

### A package per rendering back-end
We deliver a package per rendering back-end `VL.Video.MediaFoundation.Stride` and `VL.Video.MediaFoundation.Skia` each containing the respective `VideoPlayer` node. With ` (Back-end)` as version or using a sub-category?

`+` Each package comes with the dependencies it actually needs - smaller footprint on export
`-` Might be confusing for beginners as the names Skia or Stride are probably meaningless - but so would be MediaFoundation

**Comment by tebjan** [comment]
Since we should deliver video playback with vvvv, I'd go with the two package approach. It could even have a common core package if it makes sense code-wise. 

It could even make sense to make these references to VL.Stride or VL.Skia because adding an extra package reference only for video playback seems cumbersome. Skia and Stride should both have video playback out of the box...

**Comment by azeno** [comment]
Ohh I didn't think of that. Good idea to reference the back-end specific ones by default from the main drawing libraries itself. That removes the issue of having to know the package name. And yes, a common package will make a lot of sense. The back-end specific code is tiny compared to the rest.

**Comment by azeno** [comment]
Hmm, on a second thought I'm not sure that idea works out as we'd introduce a cyclic dependency. For example VL.Skia -> VL.Video... -> VL.Skia? We'd need to add a VL.Skia.Core package to break the dependency cycle.

**Comment by tebjan** [comment]
> Or place the specific nodes in the VL.Stride and VL.Skia packages directly and only reference the VL.Video.MediaFoundation.Core package.

that sounds easy enough. the only issue could be that the Stride and Skia packages would need an update when something was fixed or changed on video playback. a core and two specific packages might be the cleanest in terms of references, but cumbersome to use...

**Comment by joreg** [comment]
"the Stride and Skia packages would need an update when something was fixed or changed on video playback"

but until the parent packs are updated, ppl could still update the video packs individually. which makes the requirement to update the parents less pressing..

**Comment by joreg** [comment]
how much does an un-need mediafoundation reference contribute to an export?

als MF is windows only.. how does that influence this discussion?

**Comment by azeno** [comment]
The `SharpDX.MediaFoundation.dll` is around 500kb

And MF Windows only - true. But currenty we're hiding that fact in other packages as well. The whole thing should probably be called `VL.Video` and `VL.Video.MediaFoundation` would be something which gets plugged-in when running on Windows?

**Comment by joreg** [comment]
yes, thats for readability. i thought all VL.Stride sub-libs would best start with VL.Stride and then .Sub, for findability and sortability. 

and VL.Video.MediaFoundation would be renamed to VL.MediaFoundation, because it's shorter and more obscure, which is good, because nobody should use it directly.

**Comment by joreg** [comment]
yes, thats for readability. i thought all VL.Stride sub-libs would best start with VL.Stride and then .Sub, for findability and sortability. 

and VL.Video.MediaFoundation would be renamed to VL.MediaFoundation, because it's shorter and more obscure, which is good, because nobody should use it directly.

**Comment by azeno** [comment]
Just to be on the same page - that naming scheme wouldn't be possible for anyone else, right? We as the maintainers of the `VL.Stride.*` space wouldn't be very happy if someone else uploaded a package with that name. So anyone else would need to go with the other naming scheme - will this fall on our heads at some point?

**Comment by joreg** [comment]
"We as the maintainers of the VL.Stride.* space wouldn't be very happy if someone else uploaded a package with that name." but could we even prevent that? 

"will this fall on our heads at some point?" wouldn't we have to control the whole VL.* namespace to prevent this?

i think nuget even has a means to reserve a namespace, but i haven't looked to deep into this and didn't think about it in this particular case.

**Comment by azeno** [comment]
I just updated my post above to add our latest thoughts.

**Comment by azeno** [comment]
Maybe it helps for the discussion to bring in other examples like https://github.com/vvvv/VL.CEF - it currently uses option (1) and suffers from the same issues. In that case option (3) would be a no-go because the CEF dependency is huge (>100mb). So only (2) and (2a) would be viable in that case.

**Comment by joreg** [comment]
we've decided to leave this as is for 2021.4 but there is a new idea (to be picked up after 2021.4):
- add GPU support to IImage interface 
- register conversions for IImage -> SkImage (in VL.Skia) and IImage -> Texture (in VL.Stride) 
- that should hopefully simplify the whole discussion
